lib/libc/common/CMakeLists.txt

```diff
@@ -6,3 +6,6 @@ zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_ABORT source/stdlib/abort.c)
 zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_TIME source/time/time.c)
 zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_MALLOC source/stdlib/malloc.c)
 zephyr_library_sources_ifdef(CONFIG_COMMON_LIBC_STRNLEN source/string/strnlen.c)
+
+# Prevent compiler from optimizing stdlib functions into infinite recursive calls
+zephyr_library_cc_option(-fno-builtin)
```

Comment:
I build code like this with only the necessary builtins disabled. That allows the compiler to continue to perform optimizations and checking for other functions. In this case, adding `-fno-builtin-malloc -fno-builtin-free` would be a more specific change and will solve the problem.

Reply:
I agree that would fix this issue. My concern would be that after the next GCC release, this starts occurring with `strlen` and the other functions in here. There is not much code in this common libc library to optimize, and every function is probably somewhat susceptible to this issue

Reply:
yeah, it's a balance -- using the big hammer (`-fno-builtin`) gets rid of all checking and optimizations for all functions the compiler knows about, which makes code worse (both in performance and in correctness). Using specific options means you might miss one and cause recursion issues like this. On balance, for picolibc, I decided that I was willing to accept the latter risk as it's quite easy to test for while the former is essentially un-testable. Of course, that relies on having basic tests for each of these functions to make sure they don't recurse.

Reply:
Fair enough, I have modified the PR to use `-fno-builtin-malloc`

Reply:
Looks like we should add a bug to add tests for each one of these functions someday?